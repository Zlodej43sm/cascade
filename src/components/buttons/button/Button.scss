@use '../../../styles/colors';
@use '../../../styles/mixins';
@use '../../../styles/sizes';
@use '../../../styles/typography';

.button {
  @include mixins.resetButtonStyles;
  padding: sizes.$base sizes.$base-2;

  &[disabled] {
    opacity: .7;
    pointer-events: none;
  }
}

.primary {
  background-color: colors.$cascade-secondary;
  color: colors.$base-lightest;
}

.secondary {
  background-color: colors.$grey-10;
  color: colors.$base-grey;
}

.default {
  @include typography.button;
  border-radius: 10px;
}

.compact {
  @include typography.button-compact;
  border-radius: 7px;
  min-width: sizes.$button-compact-min-width;

  .primary {
    color: colors.$grey-0;
  }
}
