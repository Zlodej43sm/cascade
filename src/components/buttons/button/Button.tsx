import React from 'react';
import clsx from 'clsx';

import styles from './Button.scss';

export enum ButtonTypes {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ButtonSizes {
  Default = 'default',
  Compact = 'compact',
}

export type ButtonProps = Pick<React.ComponentProps<'button'>, 'type' | 'disabled' | 'onClick' | 'children'> & {
  variant?: ButtonTypes
  size?: ButtonSizes
};

function Button({
  children,
  type = 'button',
  variant = ButtonTypes.Primary,
  disabled = false,
  size = ButtonSizes.Default,
  onClick,
}: ButtonProps) {
  const className = clsx(styles.button, styles[variant], styles[size]);
  return (
    <button className={className} onClick={onClick} type={type} data-testid="button" disabled={disabled}>
      {children}
    </button>
  );
}

export default Button;
