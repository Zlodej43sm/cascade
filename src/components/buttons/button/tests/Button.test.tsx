import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import { buttonArguments } from '../__mocks__/Button.mock';
import Button, { ButtonSizes } from '../Button';

describe('Button', () => {
  it('Should match snapshot', () => {
    const { asFragment } = render(<Button {...buttonArguments.defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('Should call click handler', () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>No extra props</Button>);
    fireEvent.click(screen.getByTestId('button'));
    expect(onClick).toHaveBeenCalled();
  });
  it('Should NOT handle click on button disabled', () => {
    const onClick = jest.fn();
    render(<Button {...buttonArguments.defaultDisabledProps} onClick={onClick} />);
    fireEvent.click(screen.getByTestId('button'));
    expect(onClick).not.toHaveBeenCalled();
  });
  it('Should have compact class', () => {
    const { getByTestId } = render(<Button {...buttonArguments.defaultCompactProps} />);
    expect(getByTestId('button')).toHaveClass(ButtonSizes.Compact);
  });
  it('Should have type - submit', () => {
    const submitType = 'submit';
    const { getByTestId } = render(<Button type={submitType} {...buttonArguments.defaultProps} />);
    expect(getByTestId('button')).toHaveAttribute('type', submitType);
  });
});
