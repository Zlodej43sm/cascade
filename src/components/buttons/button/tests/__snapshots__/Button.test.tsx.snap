// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="button primary"
    data-testid="button"
    type="button"
  >
    Button primary
  </button>
</DocumentFragment>
`;
